In [1]:
import pandas as pd
import statsmodels.api as sm
import numpy as np
import matplotlib.pyplot as plt
from linearmodels.iv import IV2SLS

In [2]:
# Exercise 1
wage2 = pd.read_stata("./stata/WAGE2.DTA")
X = sm.add_constant(wage2[["sibs"]])
model = sm.OLS(wage2.lwage, X, missing="drop").fit()
model.summary()

/home/systemchalk/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  lwage   R-squared:                       0.023
Model:                            OLS   Adj. R-squared:                  0.022
Method:                 Least Squares   F-statistic:                     22.31
Date:                Wed, 23 Mar 2022   Prob (F-statistic):           2.68e-06
Time:                        02:20:12   Log-Likelihood:                -506.59
No. Observations:                 935   AIC:                             1017.
Df Residuals:                     933   BIC:                             1027.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          6.8611      0.022    310.771      0.000       6.818       6.904
sibs          -0.0279      0.006     -4.723      0.000      -0.039      -0.016
==============================================================================
Omnibus:                       22.562   Durbin-Watson:                   1.726
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               26.529
Skew:                          -0.306   Prob(JB):                     1.74e-06
Kurtosis:                       3.553   Cond. No.                         6.33
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [3]:
X = sm.add_constant(wage2[["brthord"]])
model = sm.OLS(wage2.educ, X, missing="drop").fit()
model.summary()

/home/systemchalk/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   educ   R-squared:                       0.042
Model:                            OLS   Adj. R-squared:                  0.041
Method:                 Least Squares   F-statistic:                     37.29
Date:                Wed, 23 Mar 2022   Prob (F-statistic):           1.55e-09
Time:                        02:20:13   Log-Likelihood:                -1861.9
No. Observations:                 852   AIC:                             3728.
Df Residuals:                     850   BIC:                             3737.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         14.1494      0.129    109.962      0.000      13.897      14.402
brthord       -0.2826      0.046     -6.106      0.000      -0.373      -0.192
==============================================================================
Omnibus:                       77.090   Durbin-Watson:                   1.784
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               53.992
Skew:                           0.506   Prob(JB):                     1.89e-12
Kurtosis:                       2.294   Cond. No.                         5.28
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [4]:
IV2SLS(wage2.lwage, np.ones((wage2.shape[0], 1)), wage2.educ, wage2.brthord).fit(cov_type="unadjusted")

/home/systemchalk/anaconda3/lib/python3.9/site-packages/linearmodels/shared/exceptions.py:35: MissingValueWarning: 
Inputs contain missing values. Dropping rows with missing observations.
  warnings.warn(missing_value_warning_msg, MissingValueWarning)


IV-2SLS Estimation Summary                          
==============================================================================
Dep. Variable:                  lwage   R-squared:                     -0.0286
Estimator:                    IV-2SLS   Adj. R-squared:                -0.0298
No. Observations:                 852   F-statistic:                    16.667
Date:                Wed, Mar 23 2022   P-value (F-stat)                0.0000
Time:                        02:20:13   Distribution:                  chi2(1)
Cov. Estimator:            unadjusted                                         
                                                                              
                             Parameter Estimates                              
==============================================================================
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
exog           5.0304     0.4324     11.633     0.0000      4.1828      5.8780
educ           0.1306     0.0320     4.0825     0.0000      0.0679      0.1934
==============================================================================

Endogenous: educ
Instruments: brthord
Unadjusted Covariance (Homoskedastic)
Debiased: False
IVResults, id: 0x7fe3fc116f40

In [5]:
X = sm.add_constant(wage2[["sibs", "brthord"]])
model = sm.OLS(wage2.educ, X, missing="drop").fit()
# We need the fitted values for part vi (first stage results)
educ_hat = model.fittedvalues
model.summary()

/home/systemchalk/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   educ   R-squared:                       0.058
Model:                            OLS   Adj. R-squared:                  0.056
Method:                 Least Squares   F-statistic:                     26.29
Date:                Wed, 23 Mar 2022   Prob (F-statistic):           8.33e-12
Time:                        02:20:13   Log-Likelihood:                -1854.6
No. Observations:                 852   AIC:                             3715.
Df Residuals:                     849   BIC:                             3729.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         14.2965      0.133    107.260      0.000      14.035      14.558
sibs          -0.1529      0.040     -3.834      0.000      -0.231      -0.075
brthord       -0.1527      0.057     -2.675      0.008      -0.265      -0.041
==============================================================================
Omnibus:                       69.406   Durbin-Watson:                   1.795
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               51.104
Skew:                           0.497   Prob(JB):                     8.00e-12
Kurtosis:                       2.329   Cond. No.                         8.46
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [6]:
X = sm.add_constant(wage2[["sibs"]])
IV2SLS(wage2.lwage, X, wage2.educ, wage2.brthord).fit(cov_type="unadjusted")

/home/systemchalk/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)
/home/systemchalk/anaconda3/lib/python3.9/site-packages/linearmodels/shared/exceptions.py:35: MissingValueWarning: 
Inputs contain missing values. Dropping rows with missing observations.
  warnings.warn(missing_value_warning_msg, MissingValueWarning)


IV-2SLS Estimation Summary                          
==============================================================================
Dep. Variable:                  lwage   R-squared:                     -0.0543
Estimator:                    IV-2SLS   Adj. R-squared:                -0.0568
No. Observations:                 852   F-statistic:                    21.871
Date:                Wed, Mar 23 2022   P-value (F-stat)                0.0000
Time:                        02:20:13   Distribution:                  chi2(2)
Cov. Estimator:            unadjusted                                         
                                                                              
                             Parameter Estimates                              
==============================================================================
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
const          4.9385     1.0538     4.6863     0.0000      2.8731      7.0040
sibs           0.0021     0.0173     0.1217     0.9031     -0.0319      0.0361
educ           0.1370     0.0745     1.8376     0.0661     -0.0091      0.2831
==============================================================================

Endogenous: educ
Instruments: brthord
Unadjusted Covariance (Homoskedastic)
Debiased: False
IVResults, id: 0x7fe3fc151b20

In [7]:
educ_hat.corr(wage2.sibs)

-0.929481763991843

C1.i When plugging in sibs in as a proxy for education, the coefficent is negative, sufficient to show it is not the same as the IV (0.122, from the text). Specifically this regression says that, without controlling for other factors, an additional sibling is associated with a little under 3% less in wages (though we shouldn't place too much emphasis on this as sibs is almost certianly correlated with other factors)

C1.ii One intuitive reason why education might be negatively associated with birth order is simply the matter of resource constraints (older children may have money set aside for education while younger children may not). The regression shows a statistically significant negative relationship between education and birth order, with each new position in the rank corresponding to a little over a quarter of a year less in education.

C1.iii Results reported above. When using birth order as an instrument for education, we see a larger coefficient for education.

C1.iv We are assuming that $\pi_2 \neq 0$ (that is, brthord isn't zero). The regression allows us to reject the null hypothesis that this value is zero, and so the identification assumption holds.

C1.v Results above. The standard error is considerably higher than the one estimated previously (0.0745 vs. 0.0320), enough to make it lose significance. Sibs has a small standard error, but also a small coefficient, meaning we have no evidence that it is significantly different from zero.

C1.vi There is a strong negative correlation between the fitted values of educ and sibs (-0.93). Recalling from section 3 that multicollinearity is a problem for 2SLS, we have an explanation for our large standard errors.